toolchain/driver/driver.cpp

```diff
@@ -22,286 +25,481 @@
+  void Build(Args::CommandBuilder& b) {
+    b.AddStringPositionalArg(
+        {
+            .name = "FILE",
+            .help = R"""(
+The input Carbon source file to compile.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.Required(true);
+          arg_b.Set(&input_file_name);
+        });
+
+    b.AddOneOfOption(
+        {
+            .name = "phase",
+            .help = R"""(
+Selects the compilation phase to run. These phases are always run in sequence,
+so every phase before the one selected will also be run. The default is to
+compile, lower, and generate machine code.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.SetOneOf(
+              {
+                  arg_b.OneOfValue("tokenize", Phase::Tokenize),
+                  arg_b.OneOfValue("parse", Phase::Parse),
+                  arg_b.OneOfValue("syntax", Phase::Syntax),
+                  arg_b.OneOfValue("lower", Phase::Lower),
+                  arg_b.OneOfValue("codegen", Phase::CodeGen).Default(true),
+              },
+              &phase);
+        });
+
+    b.AddStringOption(
+        {
+            .name = "output",
+            .value_name = "FILE",
+            .help = R"""(
+The output filename for codegen.
+
+When this is a file name, either textual assembly or a binary object will be
+written to it based on the flag `--asm-output`. The default is to write a binary
+object file.
+
+This can also be the string `-` which will write the output to stdout. In that
+case, the flag `--asm-output` is ignored and the output defaults to textual
+assembly. Binary object output can be forced by enabling `--force-obj-output`.
+)""",
+        },
+        [&](auto& arg_b) { arg_b.Set(&output_file_name); });
+
+    b.AddStringOption(
+        {
+            .name = "target",
+            .help = R"""(
+Select a target platform. Uses the LLVM target syntax, often called a "triple"
+(despite not actually being a triple in many common cases).
+
+This corresponds to the `target` flag to Clang and accepts the same strings
+documented there:
+https://clang.llvm.org/docs/CrossCompilation.html#target-triple
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.Default(host);
+          arg_b.Set(&target);
+        });
+
+    b.AddFlag(
+        {
+            .name = "asm-output",
+            .help = R"""(
+Write textual assembly rather than a binary object file to the code generation
+output.
+
+This flag only applies when writing to a file. When writing to stdout, the
+default is textual assembly and this flag is ignored.
+)""",
+        },
+        [&](auto& arg_b) { arg_b.Set(&asm_output); });
+
+    b.AddFlag(
+        {
+            .name = "force-obj-output",
+            .help = R"""(
+Force writing a binary object file, even when writing to stdout.
+
+This flag is only used when the code generation output file is set to stdout. In
+that case, it will override the forced default of textual assembly and output a
+binary output.
```

Comment:
Why should we emit the object code to stdout? It is a machine code and so it would print out some gibberish in the stdout which would be impossible to comprehend. Instead should we not print the assembly to stdout and force the user to provide an output file if they want the object code?

Reply:
While unusual, there are times when this is useful for using Unix pipes to process the object code. And it is something that existing compilers do allow -- they don't even use a scary looking flag for it, it is the *default* for those compilers:

```
> clang -c -o - example.c | llvm-objdump --disassemble -

<stdin>:        file format mach-o arm64

Disassembly of section __TEXT,__text:

0000000000000000 <ltmp0>:
       0: d10043ff      sub     sp, sp, #0x10
       4: 52800000      mov     w0, #0x0
       8: b9000fff      str     wzr, [sp, #0xc]
       c: 910043ff      add     sp, sp, #0x10
      10: d65f03c0      ret
```

I think this isn't a good default, but it seemed reasonable to preserve a way of achieving this where it is desired, and just make sure it is explicitly requested by the user.

Does that make sense?

Reply:
Thank you for the detailed explanation and I know that compilers allow that and this can be kept as an option for the scenario you have explained above. However, I don't think that it should be a default only because other compilers are doing it. The user can always dump the object code to a file and disassemble the file using the same command.

Reply:
If it required lots of additional logic, I think I would agree it would best be omitted. But because it is easily supported, it seems better to make things somewhat well rounded here, and it seems useful to have a clear mapping from existing compiler behavior.

But regardless, this is really a discussion for the toolchain-side of this PR. Going to resolve this and other comment threads here, and once this lands and I update the toolchain PR, that's maybe a good place to continue any discussion here.